### DISCUSSION TITLE
zig build 能不能两个产物有个前后关系

### DISCUSSION THREAD

**Discussion by Guozhanxin** [zigcc/forum#156] [General]
我有个项目会包括嵌入式和上位机两个部分。然后上位机的部分希望能把嵌入式生成的产物嵌入进来。用了@embedFile。不过目前第一次编译会报错，因为同时编译嵌入式端的产物还没编译出来。
![0efa8d7a276e2b870bab81569bc4eb4](https://github.com/user-attachments/assets/9646230c-8db1-45d5-84c7-9aba2bcd0383)

https://github.com/puppy-rtos/stm32-zboot/blob/92047c41c2c3ef8f5df77a06f4adff2eeed929c8/build.zig#L49-L62

**Comment by zhylmzr** [comment]
`Compile.step.dependOn` 可以让一个编译单元先于另一个单元进行

**Reply by jiacai2050** [reply]
embedFile 是编译期执行的，dependOn 还不行。可以试试在 comptime 中检查对应文件是否存在，不存在就调用一下，进行编译。

**Reply by Guozhanxin** [reply]
解决了，我把两个程序的编译顺序做了依赖，第二个程序的编译依赖第一个程序生成bin文件的Step。
https://github.com/puppy-rtos/stm32-zboot/blob/a575c7e563f70d8a7bd9a816cf0f5531a76dba37/build.zig#L56-L61

**Reply by jiacai2050** [reply]
那你 embedFile 怎么处理的呢？

**Reply by vkensou** [reply]
似乎dependOn设置好就行了。第一个编完再编的第二个

**Reply by jiacai2050** [reply]
不太可能，embedFile 是编译期执行的，而 dependOn 是运行时的依赖，有些说不通。

**Reply by xihale** [reply]
dependOn 是 build.zig 的运行时依赖吧，按理说加了这个之后跑完前面的编译任务才会开始编译下一步的程序吧？
